text/0021-state-pruning-and-fast-syncing.md

```diff
@@ -0,0 +1,240 @@
+1. For each block, visit all of its dependencies (deploys, finality signatures) and add them to an "objects to keep" set.
+1. Visit each block, deploy and finality signature, remove them if not inside the "objects to keep" set.
+
+In a manner similar to the trie store, we can employ a mark-and-sweep version of this pruning algorithm as well.
+
+This again reduces the problem on selecting which blocks to keep.
+
+### Pruning summary
+
+Both trie-store as well as storage can be trimmed down by garbage collection, with savings in the latter probably being more substantial than the former.
+
+In both cases the problem can reduced to identifying which root hashes or blocks must be kept and which can be discarded. Once this information is available, garbage collection becomes straight forward.
+
+An "archival node" is a node on which garbage collection is either disabled or restricted to non-transitive state information, *i.e.* it keeps all root hashes except those created in the midst of execution.
+
+### Fast-syncing
+
+Fast syncing, while beneficial in its own right, is also of interest from a storage perspective, as being able to recover information quickly from the network or external sources influences our decisions about what information must be kept.
+
+The "fast" part of fast-syncing, as compared to the status quo, is realized from two factors, namely
+
+* **side-loading** data from outside the network through external sources and
+* **transfering** state, namely trie-store state, which is currently not transferrable at all.
+
+### Side-loading
+
+We create a common trait for Blocks, Deploys or any other transferable object that states how it can be serialized and defines a hash function for its serialized representation, as well as a key to separate the hash space. As an example, a block whose hash is `5A4A55A8CA87B584960DCB0673DC1260292991EC6D2B8F214701B5224FB141D2` would be adressable as `blockv1_5A4A55A8CA87B584960DCB0673DC1260292991EC6D2B8F214701B5224FB141D2`. The `blockv1` denotes a complete scheme, that is it also pins the serialization format. While it is not necessary to separate the hash value space to avoid collisions, the prefix allows for multiple representations, should the format of a serialization or an internal data structure change at some point. We refer to data structures that satisfy this condition as *objects* from here on out.
+
+**Note**: The defunct CEP3 proposed hashing the serialized representation of an object instead of the object itself, which makes it a lot easier for external stores to handle these, as integrity can be checked without deserializing first. However, hashing the internal object minimizes the impact on the existing codebase instead.
+
+We can then define a generic mechanism for requesting these objects from other nodes. The true value however is in being able to serialize these to static external storage.
```

Comment:
We already have a generic mechanism for requesting these objects from other nodes.  I'm not a fan of it, but it's there.

Reply:
Agreed, we should be looking at how to iterate/improve on that rather than inventing a new solution.

Reply:
> We already have a generic mechanism for requesting these objects from other nodes. I'm not a fan of it, but it's there.

The current implementation is, politely put, bad. I would remove it rather sooner than later, as it completely violates the whole reactor design and has been an issue requiring workarounds in both reactor refactoring as well as the storage component rewrite. It should have never been added in the first place and yes, I take full responsibility for that since I was the one that approved the PR that added it... Let me atone for my past sins, please! :)